Moment/MomentGenerator.cc

```diff
@@ -0,0 +1,355 @@
+        _error_msg = error.getLastMessage();
+    }
+
+    // Set is the moment calculation successful or not
+    moment_response.set_success(IsSuccess());
+
+    // Get error message if any
+    moment_response.set_message(GetErrorMessage());
+
+    return collapse_results;
+}
+
+void MomentGenerator::StopCalculation() {
+    if (_image_moments) {
+        _image_moments->StopCalculation();
+    }
+}
+
+void MomentGenerator::SetMomentAxis(const CARTA::MomentRequest& moment_request) {
+    if (moment_request.axis() == CARTA::MomentAxis::SPECTRAL) {
+        _axis = _spectral_axis;
+    } else if (moment_request.axis() == CARTA::MomentAxis::STOKES) {
+        _axis = _stokes_axis;
+    } else {
+        std::cerr << "Do not support the moment axis: " << moment_request.axis() << std::endl;
+    }
+}
+
+void MomentGenerator::SetMomentTypes(const CARTA::MomentRequest& moment_request) {
+    int moments_size(moment_request.moments_size());
+    _moments.resize(moments_size);
+    for (int i = 0; i < moments_size; ++i) {
+        _moments[i] = GetMomentMode(moment_request.moments(i));
+    }
+}
+
+void MomentGenerator::SetPixelRange(const CARTA::MomentRequest& moment_request) {
+    // Set include or exclude pixel range
+    CARTA::MomentMask moment_mask = moment_request.mask();
+    float pixel_min(moment_request.pixel_range().min());
+    float pixel_max(moment_request.pixel_range().max());
+
+    if (pixel_max < pixel_min) {
+        float tmp = pixel_max;
+        pixel_max = pixel_min;
+        pixel_min = tmp;
+    }
+
+    if (moment_mask == CARTA::MomentMask::Include) {
+        _include_pix.resize(2);
+        _include_pix[0] = pixel_min;
+        _include_pix[1] = pixel_max;
+        // Clear the exclusive array
+        _exclude_pix.resize(0);
+    } else if (moment_mask == CARTA::MomentMask::Exclude) {
+        _exclude_pix.resize(2);
+        _exclude_pix[0] = pixel_min;
+        _exclude_pix[1] = pixel_max;
+        // Clear the inclusive array
+        _include_pix.resize(0);
+    } else {
+        // Clear inclusive and exclusive array
+        _include_pix.resize(0);
+        _exclude_pix.resize(0);
+    }
+}
+
+void MomentGenerator::ResetImageMoments(const casacore::ImageRegion& image_region) {
+    // Reset the sub-image
+    _sub_image.reset(new casacore::SubImage<casacore::Float>(*_image, image_region));
+
+    casacore::LogOrigin log("MomentGenerator", "MomentGenerator", WHERE);
+    casacore::LogIO os(log);
+
+    // Make an ImageMoments object (and overwrite the output file if it already exists)
+    _image_moments.reset(new ImageMoments<casacore::Float>(casacore::SubImage<casacore::Float>(*_sub_image), os, true));
+
+    // Set moment calculation progress monitor
+    _image_moments->SetProgressMonitor(this);
+}
+
+int MomentGenerator::GetMomentMode(CARTA::Moment moment) {
+    int mode(-1);
+    switch (moment) {
+        case CARTA::Moment::MEAN_OF_THE_SPECTRUM: {
+            mode = ImageMoments<casacore::Float>::AVERAGE;
```

Comment:
don't need `mode`, just `return` the value directly. 

Also, since this is just mapping CARTA moment types to casa types, why don't we just use a map?